### DISCUSSION TITLE
Is PositionMethods startLine 0-based intentionally?

### DISCUSSION THREAD

**Discussion by dylemma** [scala/scala3#12728] [Metaprogramming] [answered]
I'm using [`PositionMethods`](https://dotty.epfl.ch/api/scala/quoted/Quotes$reflectModule$PositionMethods.html) in a macro like:

```scala
case class CallerPos(filename: String, line: Int) {
	def render = s"$filename:$line"
}
object CallerPos {
	inline given CallerPos = ${ deriveCallerPos }

	def deriveCallerPos(using quotes: Quotes): Expr[CallerPos] = {
		import quotes.reflect._
		val pos = Position.ofMacroExpansion
		val filename = Expr { pos.sourceFile.jpath.getFileName.toString }
		val line = Expr { pos.startLine }
		'{ CallerPos(${filename}, ${line}) }
	}
}
```

and I noticed that if I say e.g. `val here = implicitly[CallerPos]` on line 118 of a file, it reports line 117.

The roughly-equivalent macro in scala 2.x reports line 118:
```scala
object CallerPos {
	implicit def capture: CallerPos = macro locationMacro

	def locationMacro(c: blackbox.Context) = {
		import c.universe._

		val filename = c.enclosingPosition.source.file.name
		val line = c.enclosingPosition.line

		q"_root_.io.dylemma.spac.CallerPos($filename, $line)"
	}
}
```

I wasn't sure if this was intentional - should I just resort to adding a `+1` to my macro in Scala 3, or should I make a bug report?

**Comment by soronpo** [comment]
Yeah, I was tripped over this too. You could say that all editors are wrong and should have started with line 0 and column 0 :wink:

**Reply by som-snytt** [reply]
IDK but I hope bug. In Scala 2, `Position.line0` is for calculating, `Position.line` is the 1-based user-facing line number. I think that if everything has a pretty printer, it's not too much to ask for pretty-printed position that is 1-based.
```
  private def line0       = source.offsetToLine(point)
  private def lineOffset  = source.lineToOffset(line0)
  def line: Int           = if (hasSource) line0 + 1 else 0
```

**Comment by nicolasstucki** [comment, marked as answer]
> I wasn't sure if this was intentional - should I just resort to adding a +1 to my macro in Scala 3, or should I make a bug report?

This was intentional. The line/column indexing is 0-based. The documentation should be clearer about this.

This is the position in the file which annoyingly is not the same as in editors. We chose this variant to avoid overhead in the reflection API implementation.
